edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/IMessageConsumer.cs

```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    public interface IMessageConsumer
```

Comment:
So the pattern you have here is that all "handlers" are consumers, and they consume the message that is relevant to them?
Would it be easier (for testing and for understanding) that there is one component that figures out the "type" of the message and posts on the right channel for a particular "handler"?

Reply:
I was considering this and thought that the "figuring out" part would be a big 'if' with trying to match different patterns to see if that matches to a pattern of handler. If I walk that way, there are some options:
- the regular expression for matching (or the operation to do for matching) is in the 'smart' class (e.g. the  BridgeConnector), a whole list of them. In this case there is a handler-related string (or function) in the 'smart' class, that I didn't like, the code is not cohesive (related parts are separated)
- the regular expression (or a function for match) is in the handler class, and the 'smart' class asks for it. In this case the cohesion is better. But if I do this, it is not too far from the current solution:
- Just ask all handlers I know about (which was discovered by a marker interface of IMessageConsumer), and send them all the information, so they can handle it if they want, but if not, it is not a 'smart' class concern (and that way it is not a smart class anymore)

I think this solution is more flexible, supports the scenario when different handlers handle different aspect of the same message - maybe there is no such scenario right now, but there was before the subscription watcher class was created. Before SubscriptionWathcer, every handler took care of their own subscription changes, so multiple handlers processed the subscription change message.

Reply:
Yes, this solution is certainly most flexible, but I wonder if we need this flexibility, since we have a fixed (small) set of scenarios only, and the list is not likely to grow/change. The concerns being - now we have loops in each handler, and now we need to make sure each task/loop works correctly, etc. Also, as I said, it might be easier to make sure the match works or log it if it doesn't work. 

In any case, this is not a huge concern - it is just an opinion. I am happy to discuss with you if you like this approach better, and we can figure out.